xarray/core/indexing.py

```diff
@@ -517,7 +524,15 @@ def get_duck_array(self):
 
     def __getitem__(self, key):
         key = expanded_indexer(key, self.ndim)
-        result = self.array[self.indexer_cls(key)]
+        indexer = self.indexer_cls(key)
+
+        if isinstance(indexer, OuterIndexer):
+            result = self.array.oindex[indexer]
+        elif isinstance(indexer, VectorizedIndexer):
+            result = self.array.vindex[indexer]
+        else:
+            result = self.array[indexer]
```

Comment:
refactoring this to a helper function would be nice I think, since it's duplicated down below.